### ISSUE TITLE
Allow statically typed enums and const

### ISSUE THREAD

**Issue by Shadowblitz16** [godotengine/godot-proposals#1877]
**Describe the project you are working on:**
Networking Game

**Describe the problem or limitation you are having in your project:**
I Came across this issue when messing around in code but you can't actually return  enums 

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**
Allow for the following..

```gdscript
enum ENUM {
	TEST = 0
}
const CONST = {
	TEST = 0
}

func test1(test:CONST)->CONST:
	pass

func test2(test:ENUM)->ENUM:
	pass
```

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**
Basicly functions and variables could have enums as their static type

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
yes and nope

**Is there a reason why this should be core and not an add-on in the asset library?:**
It supports more static typing

**Comment by Shadowblitz16** [comment]
Note this would have to only support const enums not const vars

**Comment by Shadowblitz16** [comment]
[related](https://github.com/godotengine/godot-proposals/issues/1882) and basically the same issue. 
the only difference is this one has more info on enum and const

**Comment by Shadowblitz16** [comment]
reopening this

**Comment by jonbonazza** [comment]
Enums and const (in this case) are just ints. So just return an int.

I agree enums should prpbably be treated as distinct types in gdscript 2.0. not sure what your thoughts are here @vnen

**Comment by Shadowblitz16** [comment]
const can basically be enums though

if they weren't they only valid syntax for them would be..
```
const var test = 2
``

**Comment by CedNaru** [comment]
The new GDscript in Godot 4.0 already lets you use Enum as proper types right now.

[Related article](https://godotengine.org/article/gdscript-progress-report-type-checking-back)

**Comment by vnen** [comment]
> ```gdscript
> const CONST = {
> 	TEST = 0
> }
> ```

This is just a dictionary assigned to a constant. It cannot really be a "type". The dictionary can have any arbitrary structure, so there's not a sane way to validate this (even if it's just a compile time check).

For enums, you can already use them as type hints in Godot 4. They are treated as integer in general but when possible they validate to the actual enum (so if you mismatch enum types it'll give an error).

**Comment by akien-mga** [comment]
Closing as this is already implemented in Godot 4.0 as per above comments.

**Comment by Shadowblitz16** [comment]
@vnen you know what I actully didn't realize 
```const test = {

}```

Was a dictionary. I ned to go update my other issue now.

also about it already being implimented. Thats good but can enums be consts
